ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+)
+
+const (
+	TaskManifestMessageName = "TaskManifestMessage"
+)
+
+// TaskComparer gets and compares running tasks on an instance to those in the ACS manifest.
+// It should be implemented by an underlying struct that has access to such data.
+type TaskComparer interface {
+	CompareRunningTasksOnInstanceWithManifest(*ecsacs.TaskManifestMessage) ([]*ecsacs.TaskIdentifier, error)
+}
+
+// SequenceNumberAccessor is used to get and set state for the current container
+// instance. It sets and gets the latest sequence number that corresponds to the most
+// recent task manifest that the task manifest responder has processed it alo sets the
+// latest message id
+type SequenceNumberAccessor interface {
+	GetLatestSequenceNumber() int64
+	SetLatestSequenceNumber(message *ecsacs.TaskManifestMessage) error
+}
+
+// taskManifestResponder responds to task manifest messages from ACS. It processes the
+// message and, if the message isn't stale, creates a list of stop candidates, which are
+// tasks that are running on the instance but are tasks that ACS says should not be running.
+// The taskManifestResponder will ack the task manifest message and send ACS a
+// TaskStopVerificationMessage if there are stop candidates.
+type taskManifestResponder struct {
+	taskComparer      TaskComparer
+	snAccessor        SequenceNumberAccessor
+	messageIDAccessor ManifestMessageIDAccessor
+	respond           wsclient.RespondFunc
+	metricsFactory    metrics.EntryFactory
+}
+
+// NewTaskManifestResponder returns an instance of the taskManifestResponder struct.
+func NewTaskManifestResponder(taskComparer TaskComparer, snAccessor SequenceNumberAccessor, messageIDAccessor ManifestMessageIDAccessor,
+	metricsFactory metrics.EntryFactory, responseSender wsclient.RespondFunc) *taskManifestResponder {
+	r := &taskManifestResponder{
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	logger.Debug(fmt.Sprintf("Processing %s with MessageId %s", TaskManifestMessageName, aws.StringValue(message.MessageId)))
```

Comment:
nit: Prefer to use `logger.Fields` and and `field.MessageID` in this log message.

Reply:
ack